x-pack/plugins/rule_registry/server/rule_data_client/index.ts

```diff
@@ -116,29 +117,22 @@ export class RuleDataClient implements IRuleDataClient {
 
     const clusterClient = await this.getClusterClient();
 
-    const { body: aliasExists } = await clusterClient.indices.existsAlias({
-      name: alias,
-    });
-
     const concreteIndexName = `${alias}-000001`;
-
-    if (!aliasExists) {
-      try {
-        await clusterClient.indices.create({
-          index: concreteIndexName,
-          body: {
-            aliases: {
-              [alias]: {
-                is_write_index: true,
-              },
+    try {
+      await clusterClient.indices.create({
+        index: concreteIndexName,
+        body: {
+          aliases: {
+            [alias]: {
+              is_write_index: true,
             },
           },
-        });
-      } catch (err) {
-        // something might have created the index already, that sounds OK
-        if (err?.meta?.body?.error?.type !== 'resource_already_exists_exception') {
-          throw err;
-        }
+        },
+      });
+    } catch (err) {
+      // something might have created the index already, that sounds OK
```

Comment:
What if the index already exists, but isn't set as the write index for the alias? Wouldn't we have to add the alias here?

Reply:
Is it possible to get into a scenario where the index exists but the alias has no write target, other than by an admin manually creating the index but creating it incorrectly? My assumption was that this function is the only place where the `.alerts` index can be created so we can rely on it being correctly associated with the alias as the write index if it exists. I was under the impression that the error suppression here was to avoid throwing an error if multiple rule executors attempt to create the index at the same time.

Perhaps if we get a `resource_already_exists_exception` we could fetch the index settings to ensure that whatever created the index created it correctly?

Reply:
I can't think of a way this code would produce inconsistent alias settings, but I've seen many situations in which backup scripts have not correctly restored all aliases. We might not be able to automatically fix all such situations, but we should try to to make the error message as useful as possible IMHO.

What if we split setting the write index for an existing index into a separate function and call it or `createWriteTargetIfNeeded` depending on the original error message (`index_not_found_exception` or `illegal_argument_exceptions` + alias check)?

Reply:
For both `index_not_found_exception` and `illegal_argument_exception` we will want to be able to create a new index to be the write target. Can we punt on adding the alias to existing indices for now and just log the errors? For the security solution use case, we need to be able to create a write index when the alias already exists and has no write index, but we don't need to support adding the alias to existing AAD indices.

Going back to the original question
> What if the index already exists, but isn't set as the write index for the alias? Wouldn't we have to add the alias here?

I think not - if the index exists but isn't set as the write index, something has gone wrong but I don't think we have enough information to safely attempt to automatically fix it. If this happens due to a scenario like backup scripts not restoring everything correctly, then (with the latest changes I pushed up) the initial bulk index would fail, we'd call `createWriteTargetIfNeeded`, and it would find the existing AAD indices even if they aren't aliased correctly so it would not modify the alias or make any indices. The writer would then attempt to write again, fail again, and throw the error up to the rule executor to be logged and maybe displayed in the UI. At that point the error should be available to bubble up to users and tell them that their alias has no write index. Does that sound reasonable?

Re: automatically fixing things, I'm imagining a case where a user restores `.alerts-000002` and `.alerts-000003` from backups but accidentally sets neither as the write index. If we optimistically create `.alerts-000001` as the new write index and set it up with ILM, then everything works fine for a while but eventually when ES attempts to rollover to `.alerts-000002` the rollover fails and would require manual intervention to get back to normal. To me, it seems worse to have the system seem to work normally for a while and then fail much later since it would be harder to figure out the root cause (it's easier to see the correlation between restoring backups and failing to write alerts if there's not a long time between the restore and when alert writes started failing).

Reply:
good points :+1: showing a relatively accurate error to the user should be enough for now